### DISCUSSION TITLE
Calling an operation on a device

### DISCUSSION THREAD

**Discussion by dmocek** [edgexfoundry/community#51] [General]
I asked this question quite a while ago.  I'm asking again in case something has changed that I'm not aware of.

Can I call a function on a device using Device Commands?  For example, if my device has a reset() function on it, can I call that reset() function?

The previous answer was No.  To call reset(), I'd have to represent the reset() function as a Device Resource.  For example creating a 'reset' device resource as type boolean with the expectation that setting it to 'true' will invoke the reset.  This isn't really making a function call on a device.  I consider this a workaround.

**Comment by cloudxxx8** [comment]
We discussed before and there is a design for this, but we haven't implemented.
https://github.com/edgexfoundry/edgex-docs/pull/644
You can leave your comments in this design

The current workaround is to use a normal Set command to achieve that, but the developers and users need to know the contest.  For example, to start the streaming in a USB camera:
https://github.com/edgexfoundry/device-usb-camera/blob/ad81b6791089a5483c30c2e43f91131867ddea47/cmd/res/profiles/general.usb.camera.yaml#L68-L74

**Comment by dmocek** [comment]
OK.  This also means it will be more difficult to support operations that take one or more parameters.

**Reply by cloudxxx8** [reply]
For the current version, it requires to define multiple device resources with different attributes (parameters) to simulate the operation with different parameters.
Or, you can pass query parameters to the Command API of Device Services
https://docs.edgexfoundry.org/2.3/design/adr/device-service/0011-DeviceService-Rest-API/#query-parameters
